### PR TITLE
Fix/consolidate validate pfd ino

### PR DESCRIPTION
**Pull request by cgoesche**
Addresses this issue https://github.com/util-linux/util-linux/issues/3837#issuecomment-3510910661

### CODE REVIEW

File: include/pidfd-utils.h [hunk 1]
```diff
@@ -94,4 +94,7 @@ static inline int pidfd_getfd(int pidfd __attribute__((unused)),
 }
 #endif
 
+extern int ul_get_valid_pidfd(pid_t pid, ino_t pfd_ino);
```
Comment (t-8ch) [inline include/pidfd-utils.h]: No need for `extern`.

File: include/pidfd-utils.h [hunk 2]
```diff
@@ -94,4 +94,7 @@ static inline int pidfd_getfd(int pidfd __attribute__((unused)),
...
+extern int validate_pfd_ino(int pfd, ino_t pfd_ino);
```
Comment (t-8ch) [inline include/pidfd-utils.h]: `ul_` prefix? Or make it static.

File: lib/pidfd-utils.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ * Authors: Christian Goeschel Ndjomouo <cgoesc2@wgu.edu> [2025]
+ */
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <errno.h>
+
+#include "pidfd-utils.h"
+
+int validate_pfd_ino(int pfd, ino_t pfd_ino)
```
Comment (t-8ch) [inline lib/pidfd-utils.c]: This can be inlined into its only caller.

File: lib/pidfd-utils.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ * @pfd_ino: A pidfd inode number that is expected to be the
+ *           same as for the new file descriptor.
+ *
+ * Pass @pfd_ino as NULL, if the pidfd should not be validated.
+ *
+ * Return: On success, 0 is returned.
+ *         On failure, a negative number will be returned and
+ *         errno is set to EBADF if pidfd_open() fails, and 
+ *         EINVAL if @pfd_ino does not match the file descriptor's
+ *         inode number.
```
Comment (t-8ch) [inline lib/pidfd-utils.c]: I am not a fan of these return values. They hide the real problem and force the caller to enumerate the failure cases.
You could make this an `_or_err()` function which will print the proper error message directly and kill the process. This should make the callers simpler and the error messages better.

Comment (karelzak) [reply lib/pidfd-utils.c]: I think returning `-errno` (or explicitly any -Exxx number) is fine in library-like code; the caller usually cares about <0 values. 

`_or_err()` is a good solution for tool used in main() or for argv[] parsing, etc. In this case (for kill(1) and friends) it makes sense.

File: lib/pidfd-utils.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        if (pfd < 0) {
+                errno = EBADF;
+                return -1;
+        }
+
+        if (pfd_ino) {
+                rc = validate_pfd_ino(pfd, pfd_ino);
+                if (rc) {
+                        errno = EINVAL;
+                        return -1;
```
Comment (karelzak) [inline lib/pidfd-utils.c]: If we are talking about return codes, I think it's better is:
```C
rc = validate_pfd_ino(pfd, pfd_ino);
if (rc<0)
    return rc;
```
and assume errno is already set in the place where the problem has been detected (validate_pfd_ino()).

File: misc-utils/kill.c [hunk 6]
```diff
@@ -534,8 +534,13 @@ static int kill_with_timeout(const struct kill_control *ctl)
...
 
-	if ((pfd = pidfd_open(ctl->pid, 0)) < 0)
-		err(EXIT_FAILURE, _("pidfd_open() failed: %d"), ctl->pid);
+	pfd = ul_get_valid_pidfd(ctl->pid, ctl->pidfd_ino);
+	if (pfd < 0) {
+		if (errno == EBADF)
+			err(EXIT_FAILURE, _("pidfd_open() failed: %d"), ctl->pid);
+		errx(EXIT_FAILURE, _("pidfd inode %"PRIu64" not found for pid %d: %s"),
+						ctl->pidfd_ino, ctl->pid, strerror(errno));
+	}
```
Comment (karelzak) [inline misc-utils/kill.c]: If you introduce ul_get_pidfd_or_err() (as suggested by @t-8ch), you can move all the error messages into the function, keeping the callers' code simpler.

File: misc-utils/kill.c [hunk 7]
```diff
@@ -603,14 +592,13 @@ static int kill_verbose(const struct kill_control *ctl)
...
-
-			rc = validate_pfd_ino(pfd, ctl->pidfd_ino);
-			if (rc < 0)
+			pfd = ul_get_valid_pidfd(ctl->pid, ctl->pidfd_ino);
+			if (pfd < 0) {
+				if (errno == EBADF)
+					err(EXIT_FAILURE, _("pidfd_open() failed: %d"), ctl->pid);
 				errx(EXIT_FAILURE, _("pidfd inode %"PRIu64" not found for pid %d: %s"),
-								ctl->pidfd_ino, ctl->pid, strerror(-rc));
+								ctl->pidfd_ino, ctl->pid, strerror(errno));
```
Comment (karelzak) [inline misc-utils/kill.c]: If there is a usable `errno`, then you can use `err()` rather than `errx()`, and it will append the error to the output, and you do not need strerror(errno).

File: lib/pidfd-utils.c (hunk 4 above)
Comment (cgoesche) [reply lib/pidfd-utils.c]: In this case, I guess if I turn it into an `or_err()` routine, it does not necessarily need to take in a `const char *errmesg`, as there are at least two different error cases that will call `err()` differently. Otherwise the caller would need to pass two different error messages and I'm not sure if that is elegant.

Does that make sense to you ?

Comment (cgoesche) [reply lib/pidfd-utils.c]: Example:

```C
int ul_get_valid_pidfd_or_err(pid_t pid, ino_t pfd_ino, const char *errmsg)
{
        int pfd, rc = 0;
        pfd = pidfd_open(pid, 0);
        if (pfd < 0) {
                rc = -(errno);
                goto end;
        }

        if (pfd_ino)
                rc = ul_validate_pfd_ino(pfd, pfd_ino);

end:
        if (rc < 0) {
                if (errno != 0)
                        err(EXIT_FAILURE, "%s: '%d'", errmsg, pid);
                errx(EXIT_FAILURE, "%s: '%d'", errmsg, pid);
        }
        return pfd;
}
```

Caller: `pfd = ul_get_valid_pidfd_or_err(ctl->pid, ctl->pidfd_ino, _("failed to get a valid pidfd"));`

Comment (t-8ch) [reply lib/pidfd-utils.c]: Both callers use the same error messages, so you can just inline them into the new function, without having to pass them around.

Also don't test `errno` for zero like that. It should be known beforehand if it is set.

Comment (cgoesche) [reply lib/pidfd-utils.c]: Great, so something along those lines (no pun intended):

```C
int ul_get_valid_pidfd_or_err(pid_t pid, ino_t pfd_ino)
{
        int pfd, rc;

        pfd = pidfd_open(pid, 0);
        if (pfd < 0)
                err(EXIT_FAILURE, _("pidfd_open() failed"));

        if (pfd_ino) {
                errno = 0;
                rc = ul_validate_pfd_ino(pfd, pfd_ino);
                if (rc < 0 && errno != 0)
                        err(EXIT_FAILURE, _("pidfd inode %"PRIu64" not found for pid %d"),
						pfd_ino, pid);
                errx(EXIT_FAILURE, _("pidfd inode %"PRIu64" not found for pid %d"),
						pfd_ino, pid);
        }
        return pfd;
}
```

`rc` might be extraneous in this case though. Let me know if this is better.

Comment (cgoesche) [reply lib/pidfd-utils.c]: I like that the caller now just has to do `pfd = ul_get_valid_pidfd_or_err(ctl->pid, ctl->pidfd_ino);`

Comment (t-8ch) [reply lib/pidfd-utils.c]: > Great, so something along those lines (no pun intended):

Yes, mostly. But also inline `ul_validate_pfd_ino()` to get rid of the weird error checking there.

Is `fstat()` on pidfds supported on all kernels supporting pidfds? If not, we'll need to handle this, too.

Comment (cgoesche) [reply lib/pidfd-utils.c]: AFAIK, `fstat()` is possible on pidfds even before the new pidfs was introduced in 6.9, but was simply useless as they weren't unique. I could be wrong though.

Comment (karelzak) [reply lib/pidfd-utils.c]: BTW, in recent months, Benno pointed out that the `_or_err()` functions with `char *errmsg` are problematic because they always call `gettext()` even when the error message is not necessary. Our goal should be to minimize this and use generic messages (hidden in the functions) if possible.

**Comment by karelzak** [comment]
@t-8ch's note about fstat() prompted me to investigate this area.
 
A good inspiration is the systemd code. Calling fstat() for pifd is fine, but the problem arises if the inode number is a 32-bit number. Systemd addresses this issue using the `name_to_handle_at()` syscall, where file_handle.f_handle is
a 64-bit number.
 
See https://github.com/systemd/systemd/blob/main/src/basic/pidfd-util.c#L233
and https://github.com/torvalds/linux/commit/b3caba8f7a34a2bbaf45ffc6ff3a49b70afeb192 (from @brauner). 
 
Maybe we should consider this too, especially for the new `getino` command.

**Comment by brauner** [comment]
> @t-8ch's note about fstat() prompted me to investigate this area.
> 
> A good inspiration is the systemd code. Calling fstat() for pifd is fine, but the problem arises if the inode number is a 32-bit number. Systemd addresses this issue using the `name_to_handle_at()` syscall, where file_handle.f_handle is a 64-bit number.
> 
> See https://github.com/systemd/systemd/blob/main/src/basic/pidfd-util.c#L233 and [torvalds/linux@b3caba8](https://github.com/torvalds/linux/commit/b3caba8f7a34a2bbaf45ffc6ff3a49b70afeb192) (from @brauner).
> 
> Maybe we should consider this too, especially for the new `getino` command.

This is all doable with stat() + an ioctl() as well. I've written about this in the pidfs code:

```
        /*
         * On 64 bit nothing special happens. The 64bit number assigned
         * to struct pid is the inode number.
         *
         * On 32 bit the 64 bit number assigned to struct pid is split
         * into two 32 bit numbers. The lower 32 bits are used as the
         * inode number and the upper 32 bits are used as the inode
         * generation number.
         *
         * On 32 bit pidfs_ino() will return the lower 32 bit. When
         * pidfs_ino() returns zero a wrap around happened. When a
         * wraparound happens the 64 bit number will be incremented by 2
         * so inode numbering starts at 2 again.
         *
         * On 64 bit comparing two pidfds is as simple as comparing
         * inode numbers.
         *
         * When a wraparound happens on 32 bit multiple pidfds with the
         * same inode number are likely to exist (This isn't a problem
         * since before pidfs pidfds used the anonymous inode meaning
         * all pidfds had the same inode number.). Userspace can
         * reconstruct the 64 bit identifier by retrieving both the
         * inode number and the inode generation number to compare or
         * use file handles.
         */
```

So on 32 bit you'd just do:

```
struct stat st1;
__u32 gen1;

fstat(pidfd, &st1);
ioctl(pidfd, FS_IOC_GETVERSION, gen1);
```

and then use `st1.st_ino` as lower 32 bits and `gen1` as upper 32 bits. Same result.

**Comment by karelzak** [comment]
@brauner, why is `stat()+ioctl()` better than `name_to_handle_at()` (or why do the systemd guys use it?:-) ). 

What about another (non-regfile) inode numbers, like a namespace file descriptor? Is the 64-bit number always composed of st_ino and FS_IOC_GETVERSION?

**Comment by brauner** [comment]
> @brauner, why is `stat()+ioctl()` better than `name_to_handle_at()` (or why do the systemd guys use it?:-) ).

I don't think one is necessarily better than the other. The advantage of the file handle is that it's uniform for 32 and 64 bit ofc.

> 
> What about another (non-regfile) inode numbers, like a namespace file descriptor? Is the 64-bit number always composed of st_ino and FS_IOC_GETVERSION?

No. :)
Namespace file descriptors have a preceeding inode number concept so we can't use the namespace id as the inode number unfortunately. I'm pretty sure it would cause regressions if we did that. And it would break on 32 bit as you'd risk having namespaces with duplicate inode numbers breaking existing tools that rely on stat() only (As mentioned in the comment this isn't a problem for pidfs because it had no actual meaningful inode numbers initially.).

**Comment by masatake** [comment]
It seems that I should add "PIDFD.INO" column to lsfd.

File: lib/pidfd-utils.c (hunk 4 above)
Comment (cgoesche) [reply lib/pidfd-utils.c]: Good point! Making the generic error messages available for gettext with `N_("...")` is enough, no ?

**Comment by cgoesche** [comment]
So, if I understand correctly we can do the following:

1. If 64 bit, simply compare the `stat.st_ino`
2. On 32 bit, reconstruct the inode with the lower and upper bits from `stat()+ioctl()`, respectively, and then compare.

Additionally, I think we should also make sure we have a valid pidfs fd, no ?
I am thinking of something like:

```C
#define PID_FS_MAGIC		0x50494446	/* PIDFS */
...
FS_TYPE_EQUAL(...)
```

File: lib/pidfd-utils.c (hunk 4 above)
Comment (t-8ch) [reply lib/pidfd-utils.c]: > problematic because they always call gettext() even when the error message is not necessary.

Does it really matter? These are not used in libraries.

**Comment by t-8ch** [comment]
What about using `statx()` where `stx_ino` is always 64bit?

**Comment by karelzak** [comment]
> What about using `statx()` where `stx_ino` is always 64bit?

Yes. That's a good way to go (there is already statx() used in libmount, so the build system is ready). The best approach would be to make all the ino validations depend on pidof and statx, with no fallback, and without caring about 32-bit systems. 

* rename `#define USE_KILL_WITH_PIDFDINO` to `USE_PIDFD_INO_SUPPORT` and move it pidfd-utils.h
* extend the ifdef around the USE_ to also check for `defined(HAVE_STATX) `

@cgoesche see include/statfs_magic.h, seems we need  PIDFS there, and also in libmount/src/utils.c: mnt_statfs_get_fstype() :-)

File: lib/pidfd-utils.c (hunk 4 above)
Comment (karelzak) [reply lib/pidfd-utils.c]: > > problematic because they always call gettext() even when the error message is not necessary.
> 
> Does it really matter? These are not used in libraries.

It means that even if you do not need the message, executing the code triggers all the libc infrastructure. The libc design is to postpone these things until they're really needed.

File: lib/pidfd-utils.c [hunk 8]
```diff
@@ -0,0 +1,95 @@
+/*
+ * No copyright is claimed.  This code is in the public domain; do with
+ * it what you wish.
+ *
+ * Authors: Christian Goeschel Ndjomouo <cgoesc2@wgu.edu> [2025]
+ */
+#define _GNU_SOURCE 1
+
```
Comment (karelzak) [inline lib/pidfd-utils.c]: This is already in config.h (as generated by autotools/meson).

File: lib/pidfd-utils.c [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	rc = fstatfs(pidfd, &stfs);
+	if (rc < 0)
+		return 0;
+
+	return F_TYPE_EQUAL(stfs.f_type, STATFS_PIDFS_MAGIC);
+}
+
+#ifdef USE_PIDFD_INO_SUPPORT
+ino_t pidfd_get_inode(int pidfd)
```
Comment (karelzak) [inline lib/pidfd-utils.c]: I guess we need uint64_t here to make it usable on 32-bit systems. We use statx() because it doesn't use ino_t.

Comment (karelzak) [reply lib/pidfd-utils.c]: Ah ... now I'm not sure. The libc headers can already be ready; the problem is the syscall. Can you verify it? (Sorry, I'm busy with something else).

Comment (cgoesche) [reply lib/pidfd-utils.c]: No, I think you are right it's safer to use `uint64_t`, stx_ino is still a `__u64`. That means it would be better to use `uint64_t` for `pidfd_ino` as we compare the numbers (stx_ino & pidfd_ino), no ?

```C
ul_get_valid_pidfd_or_err(pid_t pid, uint64_t pidfd_ino __attribute__((__unused__)))
```

File: lib/pidfd-utils.c [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+ *           same as for the new file descriptor.
+ *
+ * Pass @pfd_ino as NULL, if the pidfd should not be validated.
+ *
+ * Return: On success, a file descriptor is returned.
+ *         On failure, err() or errx() is called to
+ *         print an error message and kill the program.
+ *
+ */
+int ul_get_valid_pidfd_or_err(pid_t pid, uint64_t pidfd_ino __attribute__((__unused__)))
```
Comment (karelzak) [inline lib/pidfd-utils.c:65]: It seems incorrect to use `__unused__` if you use it in the case of `#ifdef USE_PIDFD_INO_SUPPORT` :-)

It would be more robust to have
```
#ifdef USE_PIDFD_INO_SUPPORT
int ul_get_valid_pidfd_or_err(pid_t pid, uint64_t pidfd_ino)
#else
int ul_get_valid_pidfd_or_err(pid_t pid, uint64_t pidfd_ino __attribute__((__unused__)))
#endif
```

**Comment by karelzak** [comment]
@t-8ch please, can you also review? Thanks.

File: lib/pidfd-utils.c [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ *         On failure, err() or errx() is called to
+ *         print an error message and kill the program.
+ *
+ */
+int ul_get_valid_pidfd_or_err(pid_t pid, ino_t pfd_ino)
+{
+        int pfd, rc;
+        struct stat f;
+
+        pfd = pidfd_open(pid, 0);
```
Comment (brauner) [inline lib/pidfd-utils.c]: You could open by file handle here then you get verification and opening in one step.

**Comment by karelzak** [comment]
Merged, let's test it on users :-)
